src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+            // then the index must be somewhere on this line
+            if (lineCharacterCountIndexArray[lineNumber] > indexWithinDoc) {
+                return lineNumber;
+            }
+        }
+    }
+
+    /**
+     * Given the character offset from the beginning of the document
+     * creates an object which has the position information
+     * @param {Array<number>} lineCharacterCountIndexArray See '_createLineCharacterCountIndex'
+     * @param {number} startSearchingWithLine Line number to start search
+     * @param {number} indexWithinDoc The index of the character offset from the start of the document
+     * @return {{line: number, ch: number}} Line and character offsets
+     */
+    function _createPosFromIndex(lineCharacterCountIndexArray, startSearchingWithLine, indexWithinDoc) {
+        var lineNumber = _lineFromIndex(lineCharacterCountIndexArray, startSearchingWithLine, indexWithinDoc);
+
+        var previousLineEndingCharacterIndex = lineNumber > 0 ? lineCharacterCountIndexArray[lineNumber - 1] : 0;
+        // create a Pos with the line number and the character offset relative to the beginning of this line
+        return {line: lineNumber, ch: indexWithinDoc - previousLineEndingCharacterIndex };
+    }
+
+    /**
+     * Returns the character offset from the beginning of the document based on
+     * object properties as pos.from.line and pos.from.ch
+     * where line is the line number in the document and ch is the character offset on the line
+     * @param {Array<number>} lineCharacterCountIndexArray See '_createLineCharacterCountIndex'
+     * @param {{line: number, ch: number}} pos Object describing the position within the document
+     */
+    function _indexFromPos(lineCharacterCountIndexArray, pos) {
+        var indexAtStartOfLine = 0;
+        if (pos.line > 0) {
+            // Start with the sum of the character count at the end of previous line
+            indexAtStartOfLine = lineCharacterCountIndexArray[pos.line - 1];
+        }
+        // Add the number of characters offset from the start and return
+        return indexAtStartOfLine + pos.ch;
+    }
+
+    /**
+     * Scans entire document and callback with each match found.
+     * Uses the documentIndex to more efficiently create the position objects on found matches.
+     *
+     * @param {Array<number>} lineCharacterCountIndexArray See '_createLineCharacterCountIndex'
+     * @param {String}        documentText                 Text to scan
+     * @param {RegExp}        regex                        Regular expression used to search
+     * @param {{from: {line: number, ch: number}, to: {line: number, ch: number}}} range Area to scan for matches
+     * @param {function({line: number, ch: number}, {line: number, ch: number}, Array )} fnEachMatch Function is called with start position, end position and Regex match Array
+     */
+    function _scanDocumentUsingRegularExpression(documentIndex, documentText, regex, range, fnEachMatch) {
```

Comment:
Name `_scanDocumentUsingRegularExpression()` is excessive, it seems. Just `_scanDocument()` would do just fine since it's the only scanning function in the module.